File: src/main/kotlin/nl/pindab0ter/aoc2015/day08/Matchsticks.kt
package nl.pindab0ter.aoc2015.day08

import nl.pindab0ter.common.getInput

private val hexRegex = """(?!<\\)\\x([a-f0-9]{2})""".toRegex()

fun parse(string: String) = string
    .trim()
    .replace("""\\""", "^")
    .replace(hexRegex) { matchResult ->
        val hexValue = matchResult.groupValues[1].toInt(16)
        hexValue.toChar().toString()
    }
    .replace("""\"""", """"""")
    .replace("^", """\""")
    .replace("""^"|"$""".toRegex(), "")


fun main() {
    val input = getInput(2015, 8)

    val lines = input.lines()
    val preParseLength = lines.sumOf(String::length)

    val parsedLines = input.lines().map(::parse)
    val postParseLength = parsedLines.sumOf(String::length)

    println("Difference between character count pre and post parse: ${preParseLength - postParseLength}")
}

File: src/main/clojure/nl/pindab0ter/aoc2015/day08/matchsticks.clj
(ns nl.pindab0ter.aoc2015.day08.matchsticks
  (:require [clojure.string :as str]
            [nl.pindab0ter.common.advent-of-code :refer [get-input]]
            [nl.pindab0ter.common.collections :refer [sum]]))

(defn hex-to-char [^String s]
  (-> s second (Integer/parseInt 16) char str))

(defn unescape [^String s]
  (-> s
      (str/replace "\\\\" "^")                              ; `\\`   -> `^`
      (str/replace #"(?!<^\\)\\x([a-f0-9]{2})" hex-to-char) ; `\x45` -> `a`
      (str/replace "\\\"" "\"")                             ; `\"`   -> `"`
      (str/replace "^" "\\")                                ; `^`    -> `\`
      (str/replace #"^\"|\"$" "")))                         ; Removes first and last `"`

(defn escape [^String s]
  (-> s
      (str/escape {\\ "\\\\"
                   \" "\\\""})
      (#(str \" % \"))))

(defn -main []
  (let [lines     (str/split-lines (get-input 2015 8))
        raw       (sum (map count lines))
        unescaped (sum (map count (map unescape lines)))
        escaped   (sum (map count (map escape lines)))]
    (println "Difference between character count pre and post unescaping:" (- raw unescaped))
    (println "Difference between character count pre and post escaping:" (- escaped raw))))